### PR TITLE
Updates for openssl v3

### PR DESCRIPTION
**Pull request by timlegge**
# Description

The biggest code change to support openssl v3 was the callback function changed.  However, due to https://github.com/openssl/openssl/issues/19092 changing the pk12 password is broken in all current openssl v3.x versions released so I have a skip step in the test which adds Crypt::OpenSSL::Guess as a dependency.

Also, I exported the p12 file and recreated but I just noticed someone else may have a better command in #40.  The real trouble is that the key for the ca is not included so updating the date on the CA (which was only a year is problematic).  I had to set the date to 2007 and do:

`openssl pkcs12 -export -inkey priv.pem -in cer1.pem -certfile cer2.pem -chain -CAfile cer2.pem -out certs/test.p12
`

Fixes/addresses (If applicable) # (issue)
#40 

## Type of change

Please delete options that are not relevant.

- [x ] Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x ] My code follows the style guidelines of this project
- [ x] I have performed a self-review of my own code
- [ x] I have commented my code, particularly in hard-to-understand areas
- [x ] I have made corresponding changes to the documentation
- [ x] My changes generate no new warnings
- [x ] I have added tests that prove my fix is effective or that my feature works
- [ x] New and existing unit tests pass locally with my changes
- [ x] Any dependent changes have been merged and published in downstream modules
- [ x] Changes have been manually tested (please provide information on test platform using the fields below)

## Test / Development Platform Information

- Operating system and version
- Ubuntu 22.04.3 LTS
- Crypt::OpenSSL::PKCS12 version master
- Perl version v5.36.1
- OpenSSL version
- OpenSSL 3.0.2

Please see the issue template for a more information on provided the requested information.

### CODE REVIEW

**Comment by timlegge** [comment]
Pushed a new version of the test.p12 with the friendlyname set so it is closer to the original if that matters

**Comment by timlegge** [comment]
@GillesT78 what is 19092

**Comment by timlegge** [comment]
Oh yeah, if memory serves that bug is fixed in the next version of OpenSSL.
No released version has it at this point in time. That's why the tests for
this module had to be modified.

**Comment by timlegge** [comment]
> Hello timledgge
> 
> I don't find any information regarding the fix in the next version Do you know where I can find the fix ? in a specific branch ?
> 
> Regards Gilles

https://github.com/openssl/openssl/pull/20134 looks like it fixes the issue but I have not tested

**Comment by jonasbn** [comment]
Hi @timlegge 

I am in the process of reviewing the PR and it looks good. However I cannot get the tests to pass.

- macOS ventura 13.4.1 (c) (22F770820d)
- Perl version v5.36.1
- OpenSSL version_ 1.1.1v

**Comment by jonasbn** [comment]
@timlegge `Crypt::OpenSSL::Guess` should also be added to the: `Makefile.PL`

**Comment by timlegge** [comment]
I will take a look see if I can see anything. Can you send me a copy of the
log?

**Comment by timlegge** [comment]
@jonasbn can you run:

`perl -I blib/lib/ -I blib/arch/ t/pkcs12-string.t`

It will be easier to see what is failing

**Comment by timlegge** [comment]
> Hi @timlegge
> 
> I am in the process of reviewing the PR and it looks good. However I cannot get the tests to pass.
> 
> * macOS ventura 13.4.1 (c) (22F770820d)
> * Perl version v5.36.1
> * OpenSSL version_ 1.1.1v

I did not realise that I have bot enabled the actions on my fork.  I am able to reproduce so I will take a look.

Tim

**Comment by jonasbn** [comment]
Hi @timlegge 

Here is the output of the requested run

```
❯ carton exec -- perl -I blib/lib/ -I blib/arch/ t/pkcs12-string.t
1..29
ok 1 - use Crypt::OpenSSL::PKCS12;
ok 2
ok 3
ok 4 - utf8 string refused new_from_string()
ok 5
ok 6 - utf8 string refused in new_from_string()
# Attempting to read certificate string from file certs/test.p12
ok 7
ok 8
ok 9
ok 10
ok 11
ok 12
ok 13
ok 14
ok 15
ok 16
ok 17
ok 18 - PKCS object created
ok 19 - PEM certificate created
ok 20 - Asserting PEM key
ok 21 - Asserting mac
ok 22 - Asserting PKCS12 as string
PKCS12_newpass: testing foo
8101240320:error:0D0680A8:asn1 encoding routines:asn1_check_tlen:wrong tag:crypto/asn1/tasn_dec.c:1149:
8101240320:error:0D06C03A:asn1 encoding routines:asn1_d2i_ex_primitive:nested asn1 error:crypto/asn1/tasn_dec.c:713:
8101240320:error:0D08303A:asn1 encoding routines:asn1_template_noexp_d2i:nested asn1 error:crypto/asn1/tasn_dec.c:646:Field=salt, Type=PBEPARAM
8101240320:error:0D0C706E:asn1 encoding routines:ASN1_item_unpack:decode error:crypto/asn1/asn_pack.c:60:
8101240320:error:23080072:PKCS12 routines:PKCS12_newpass:parse error:crypto/pkcs12/p12_npas.c:50:
not ok 23 - Changing password
#   Failed test 'Changing password'
#   at t/pkcs12-string.t line 105.
#
PKCS12_verify_mac:
# Looks like your test exited with 255 just after 23.
#
```

**Comment by timlegge** [comment]
@jonasbn perfect thanks.  I suspect that the format of the file changed but I will take a closer look

**Comment by timlegge** [comment]
@jonasbn - still trying to figure this out - but I may need to create different p12 files for certain versions.  I will keep looking.  The other option is to use openssl to generate a p12 for each test as part of the test.

**Comment by timlegge** [comment]
So, the story here is that openssl has had issue with the pkcs12 "app" some changes over the versions have broken the format.  Between and across versions.

Right now I have copied the original version in to a certs/test_le_1.1.p12 and set the tests to use that file if the openssl version is less than or equal to 1.1.  That makes the tests work.  However I have a suspicion that a 1.1 version of openssl *may* not be able to create a pkcs12 file compatiable with the change password function

I will do some more testing.  It may take creating a test that creates a new pkcs12 file each time it is run to figure our where it breaks down

**Comment by timlegge** [comment]
@jonasbn I have everything working except the MacOS action.  Unfortunately nothing I try seems to fix the issue and I have no real way to figur out what going on in MacOS (I don't run it)  

I will cleanup the PR because I have been just trying different approaches.  I suspect at this point it is either a bug specific to MasOS or that specific version of openSSL.

Do you have MacOS?  If so can you test once I clean up the PR?

**Comment by jonasbn** [comment]
Hi @timlegge 

I can get the tests to pass locally.

I will make a few minor changes to the branch before merging it.

- I have commented in the unlinking of `t/scratch-csr.pem`
- I have updated the macOS GH Action, let's see if that works

**Comment by timlegge** [comment]
I'm still not sure why the GitHub action fails for that PR. I need to pull back a commit that I put on it just recently to see if that was causing it but didn't do it right. I will have to take a closer look at the versioning on that Mac github action and see if there's anything that looks suspicious.

**Comment by gregoa** [comment]
We are applying this PR in Debian (in the version from mid-August), but the two tests `t/pkcs12-string.t` and `t/pkcs12.t`are failing again against newer OpenSSL versions (3.0.11 and 3.1.3). Cf. https://bugs.debian.org/1052331

I've now updated our version of the patch to
```
    if ($major =~ /^3\./) {
        skip("OpenSSL 3.x cannot change pkcs12 passwords", 3);
```
in both `*.t`files.

**Comment by timlegge** [comment]
@jonasbn I finally realise what is happening here for the macos tester.  It likely finds and builds for openssl version 3 but the Makefile forces 1.1 I will fix and rebase in the next few days

**Comment by timlegge** [comment]
Having trouble getting Module::Install to do what I want.  Any objection to a PR to convert to Dist::Zilld based on perl-crypt-openssl-x509?

TIm

**Comment by jonasbn** [comment]
hi @timlegge 

I use Dist::Zilla everywhere else, so a PR would be most welcome

**Comment by timlegge** [comment]
@jonasbn can you approve the workflow to run.  I can still squash but I would like to see if this sorts the mac issue

**Comment by timlegge** [comment]
Ping @jonasbn .

**Comment by timlegge** [comment]
Still trying to sort out actions issues.

**Comment by jonasbn** [comment]
Hi @timlegge 

I am unable to replicate the build errors locally, so I believe this to be an issue for the CI job on GitHub.

My suspicion is that this might be related to the writing of the temporary files. I can see that we are writing to both: `t/` and `certs/`, should we make this uniform.

And the use of `unlink` in `pkcs12-from-scratch.t` is both in the top and the bottom of the file, I think we should attempt to put these in the top, so we could push the generated files as artifacts for download and inspection - let me know what you think.

Secondly we could release a TRIAL and get some feedback from CPAN-testers, perhaps they can replicate the issue.

**Comment by jonasbn** [comment]
@timlegge I believe we can remove the following lines from: `.github/workflows/macos.yml`

```yaml
      - name: install openssl
        run: |
          brew install openssl@1.1
```

The macOS runner is defaulting to OpenSSL version 3. (`3.2.1` to be exact).

What do you think?

**Comment by timlegge** [comment]
> I am unable to replicate the build errors locally, so I believe this to be an issue for the CI job on GitHub.

I am glad you weren't able to replicate it. The GitHub actions are nice but sometimes cause problems on their own. 

> My suspicion is that this might be related to the writing of the temporary files. I can see that we are writing to both: `t/` and `certs/`, should we make this uniform.
I seem to remember something around. Get up actions and temporary files. I will have to take a look 

> And the use of `unlink` in `pkcs12-from-scratch.t` is both in the top and the bottom of the file, I think we should attempt to put these in the top, so we could push the generated files as artifacts for download and inspection - let me know what you think.

Yeah, it would be nice to be able to review some of those items when there are issues 

> Secondly we could release a TRIAL and get some feedback from CPAN-testers, perhaps they can replicate the issue.
Yep, that would help a lot. 

Tim

**Comment by timlegge** [comment]
> @timlegge I believe we can remove the following lines from: `.github/workflows/macos.yml`
> 
> ```yaml
>       - name: install openssl
>         run: |
>           brew install openssl@1.1
> ```
> 
> The macOS runner is defaulting to OpenSSL version 3. (`3.2.1` to be exact).
> 
> Version 1 is deprecated, currently pointing to: `1.1.1w`
> 
> REF: https://formulae.brew.sh/formula/openssl@1.1
> 
> What do you think?

I think you are correct. When I went through I must have missed that one. Although I did specify Main for some of the installs, which we may want to go to the actual latest version.

Tim

File: dist.ini [hunk 1]
```diff
@@ -0,0 +1,108 @@
+; REF: Dist::Zilla https://metacpan.org/pod/Dist::Zilla
+name = Crypt-OpenSSL-PKCS12
+author = Dan Sully
+main_module = PKCS12.pm
+license = Perl_5
+copyright_holder = Dan Sully
+copyright_year = 2004-2021
```
Comment (jonasbn) [inline dist.ini:7]: Please bump the year

```suggestion
copyright_year = 2004-2024
```

File: .github/workflows/cygwin.yml [hunk 2]
```diff
@@ -31,22 +31,29 @@ jobs:
...
           platform: x64
-          packages: make perl gcc-core gcc-g++ pkg-config libcrypt-devel libssl-devel git curl
+          packages: make perl gcc-core gcc-g++ pkg-config libcrypt-devel libssl-devel libnsl-devel git curl
 
       - name: perl -V
         run: |
           perl -V
           gcc --version
-      - name: Run Tests
+      - name: Install Depends
```
Comment (jonasbn) [inline .github/workflows/cygwin.yml:46]: Write this out

```suggestion
      - name: Install Dependencies
```

File: .github/workflows/macos.yml [hunk 3]
```diff
@@ -19,23 +19,23 @@ jobs:
...
       - name: Set up Perl
         run: |
           brew install perl
           curl https://cpanmin.us | perl - App::cpanminus -n
           echo "/Users/runner/perl5/bin" >> $GITHUB_PATH
       - name: perl -V
         run: perl -V
-
-      - name: Run Tests
+      - name: Install Depends
```
Comment (jonasbn) [inline .github/workflows/macos.yml:32]: Write this out

```suggestion
      - name: Install Dependencies
```

File: .github/workflows/msys2-mingw.yml [hunk 4]
```diff
@@ -28,24 +28,28 @@ jobs:
...
             mingw-w64-x86_64-perl
-            msys/libcrypt-devel
             msys/openssl-devel
       - name: perl -V
         run: |
           perl -V
-          perl -V:make
-      - name: Run Tests
+          uname -a
+      - name: Install Depends
```
Comment (jonasbn) [inline .github/workflows/msys2-mingw.yml:46]: Write this out

```suggestion
      - name: Install Dependencies
```

File: .github/workflows/windows.yml [hunk 5]
```diff
@@ -33,11 +33,17 @@ jobs:
           echo "C:\strawberry\perl\bin" | Out-File -FilePath $env:GITHUB_PATH -Encoding utf8 -Append
       - name: perl -V
         run: perl -V
-
-      - name: Run Tests
+      - name: Download cpanm
+        run: |
+          cpan App::cpanminus
+      - name: Install Depends
```
Comment (jonasbn) [inline .github/workflows/windows.yml:39]: Write this out

```suggestion
      - name: Install Dependencies
```

**Review by jonasbn — CHANGES_REQUESTED** [review]
Hi @timlegge 

I have a few cosmetic suggestions. I believe I have review many of the proposed changes earlier, so I think the core functionality is in good state

I can see that you are proposing a migration from Module::Install to Dist::Zilla, so I guess we need to remove `inc/`

Let me know if you have any questions

**Comment by timlegge** [comment]
I think I fixed all the issues

**Comment by timlegge** [comment]
also fixed msys2 action

**Review by jonasbn — APPROVED** [review]
This is approved.

Thanks @timlegge for your contribution and tireless efforts

**Pull request merged**